doc/fwutil/fwutil.md

```diff
@@ -0,0 +1,441 @@
+
+## Revision
+
+| Rev | Date       | Author         | Description                       |
+|:---:|:----------:|:--------------:|:----------------------------------|
+| 0.1 | 21/08/2019 | Nazarii Hnydyn | Initial version                   |
+| 0.2 | 10/09/2019 | Nazarii Hnydyn | Review feedback and other changes |
+| 0.3 | 17/09/2019 | Nazarii Hnydyn | Align flows with the platform API |
+
+## Abbreviations
+
+| Term   | Meaning                                             |
+|:-------|:----------------------------------------------------|
+| FW     | Firmware                                            |
+| SONiC  | Software for Open Networking in the Cloud           |
+| PSU    | Power Supply Unit                                   |
+| QSFP   | Quad Small Form-factor Pluggable                    |
+| EEPROM | Electrically Erasable Programmable Read-Only Memory |
+| I2C    | Inter-Integrated Circuit                            |
+| SPI    | Serial Peripheral Interface                         |
+| JTAG   | Joint Test Action Group                             |
+| BIOS   | Basic Input/Output System                           |
+| CPLD   | Complex Programmable Logic Device                   |
+| FPGA   | Field-Programmable Gate Array                       |
+| URL    | Uniform Resource Locator                            |
+| API    | Application Programming Interface                   |
+| N/A    | Not Applicable/Not Available                        |
+
+## List of figures
+
+[Figure 1: FW utility High Level Design](#figure-1-fw-utility-high-level-design)  
+[Figure 2: Show components status flow](#figure-2-show-components-status-flow)  
+[Figure 3: FW install (non modular) flow](#figure-3-fw-install-non-modular-flow)  
+[Figure 4: FW install (modular) flow](#figure-4-fw-install-modular-flow)  
+
+## List of tables
+
+[Table 1: Event logging](#table-1-event-logging)
+
+# 1 Introduction
+
+## 1.1 Feature overview
+
+A modern network switch is a sophisticated equipment which consists of many auxiliary components  
+which are responsible for managing different subsystems (e.g., PSU/FAN/QSFP/EEPROM/THERMAL)  
+and providing necessary interfaces (e.g., I2C/SPI/JTAG).
+
+Basically these components are complex programmable logic devices with it's own HW architecture  
+and software. The most important are BIOS/CPLD/FPGA etc.
+
+It is very important to always have the latest recommended software version to improve device stability,  
+security and performance. Also, software updates can add new features and remove outdated ones.
+
+In order to make software update as simple as possible and to provide a nice user frindly  
+interface for various maintenance operations (e.g., install a new FW or query current version)  
+we might need a dedicated FW utility.
+
+## 1.2 Requirements
+
+### 1.2.1 Functionality
+
+**This feature will support the following functionality:**
+1. Manual FW installation for particular platform component
```

Comment:
Can the user schedule the reloads after the components being installed?

Reply:
Scheduling a reload is not in the scope of this design. Once the command is completed the user is being notified following the platform API implementation if reboot, power shutdown or nothing is expected from him. If schedule reboot is supported in SONiC you may do that if not this is an additional req for SONiC